Add template for fine-tuning to support function-calling

SumanthRH opened the pull request:
# What does this PR do?

Adds an example for fine-tuning on function calling data through both the Anyscale Platform and through serverless endpoints. I've also added a .gitignore file to the template repo and made some changes to the pre-commit hoooks to support larger notebooks.

kouroshHakha commented:
A few quick feedbacks: 

- Generally try to match the format of existing templates: https://github.com/anyscale/templates/blob/f939e4664d4b362b7ebdef4754ddd4332e40753c/templates/fine-tune-llm_v2/README.ipynb#L76
- Change title: Function calling fine-tuning on custom data

- Move all the imports to the first section that you want to import. Separate them into two blocks (third party imports and custom module imports)

```
import datasets
import sys
import ray.data 
```

```
from fc_utils.preprocessing import filter_func, initial_mapper, pprint_example, preprocess
```

- In each section in the code block, leave a concise in-block comment on what you are doing.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    def values():
+        return [item.value for item in PossibleMistakes]
+
+def is_match(response: dict, ground_truth: dict) -> Tuple[bool, str]:
```
Also for convention let's use the following docstrings format:

```
"""<single line description>

<more description with maybe examples (if really needed)>

Args:
    foo: explain foo line 1
        explain for line 2
    bar: explain bar (note that type is not defined here, it's defined in the signature)

Returns:
    what gets returned + (optional) what gets raised
"""

```

kouroshHakha replied:
Also the output type is `Tuple[bool, PossibleMistakes]`

SumanthRH replied:
> never use python without specific type hints (e.g. dict: Dict[str, Any] or Dict[str, np.array] or Dict[str, str] ??).

Will add. 

> Also the output type is Tuple[bool, PossibleMistakes]
Actually the output type I ended up using was a string (i passed the enum value as I can use it directly in plotting), but I can switch to the Enum class.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    """
+    Checks if the response matches the ground truth. Returns a boolean and a string indicating the mistake type if any.
+    """
+    if ground_truth["tool_calls"] is None:
```
comment each part so that the reader of the code can easily follow, eg:

```
# If ground truth is a normal response, so should be the actual response
```

Also rewrite for better readablity:

```
is_match = True
reason = ""


if x:
    is_match = False
    reason = Foo
elif y:
    is_match = False
    reason = Bar
...

return is_match, reason
```

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+
+
+def parse_and_eval(
+    parser: Union[OpenAIResponseParser, AnyscaleResponseParser],
```
Let's define a base class for these called `ResponseParser` and extend it to build these parsers?

SumanthRH replied:
Done!

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    """
+    Parse and eval loop to parse the assistant responses and evaluate them against the ground truth in the conversation.
+    Args:
+    parser: OpenAIResponseParser or AnyscaleResponseParser object
```
don't forget the indentation
+ first line should be no white-space / new line just a single line concise description.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    messages,
+    expected_responses,
+    tools=None,
+):
```
return type?

kouroshHakha commented on .pre-commit-config.yaml:
```diff
@@ -4,6 +4,7 @@ repos:
     rev: v3.2.0
     hooks:
       - id: check-added-large-files
+        args: ['--maxkb=10000']
```
what is this for?

SumanthRH replied:
This is because the current limit can be restrictive for notebooks. My notebook's size is currently 1.3MB. I believe the default limit is around 800KB or so. I made it 10MB for a lot of wiggle room.

kouroshHakha replied:
Can you add a concise but informative comment on what this flag does?

SumanthRH replied:
Done!

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
copy the Readme_files to under assets?

SumanthRH replied:
Hmm these were auto-generated from the readme conversion. Let me fix this.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    if ground_truth["tool_calls"] is None:
+        if response["tool_calls"] is None:
+            return True, ""
+        else:  # explicit else for clarity
```
remove explicit else for clarity?

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+
+    if response["tool_calls"] is None:
+        return False, PossibleMistakes.NO_FUNCTION_CALL.value
+    elif response["tool_calls"] == INCORRECT_FORMAT: # error during parsing
```
NIT: never use comments that follow code:

Dont do
```
foo = bar # setting foo to bar
```
Instead do
```
# setting foo to bar
foo = bar
```

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    for expected_tool_call, actual_tool_call in zip(ground_truth["tool_calls"], response["tool_calls"]):
+        if "name" not in actual_tool_call or "arguments" not in actual_tool_call:
+            return False, PossibleMistakes.INCORRECT_FORMAT.value
+        if expected_tool_call["name"] != actual_tool_call["name"]:
```
elif

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+            if not match:  # return right away if model output is incorrect
+                return current_conv, match, mistake_type
+        conv_id += 1 # next message
+    return current_conv, match, mistake_type
```
here return is three items but above you are returning two items. Please fix to be consistent.

SumanthRH replied:
Whoa. Fixed.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+                    current_conv[-1]["tool_call_id"] = current_conv[-2]["tool_calls"][0].id
+                else:
+                    return current_conv, False
+            if isinstance(parser, OpenAIResponseParser):
+                parsed_response = parser.get_parsed_response(current_conv, tools)
+            else:
+                parsed_response = parser.get_parsed_response(current_conv)
```
when you define the interface, design it such that it is unified:
```
parsed_response = parser.get_parsed_response(messages, tools)
```
but in the anyscale one don't use tools.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    Args:
+    parser: OpenAIResponseParser or AnyscaleResponseParser object
+    messages: list of messages in the conversation. The messages with role 'assistant' are ignored and triggers a chat completion call
+    expected_responses: list of ground truth responses
```
can you add comments / explain why is this a list and not a single assistant message?

kouroshHakha replied:
Also style nit: capitalize the first char.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    assert messages[0]["role"] == "system", "First message must be from system"
+    match = True
+    current_conv = []
+    assist_id = 0
+    conv_id = 0
```
nit: for index use idx or index not id. also name them: `cur_message_idx` and `cur_expected_idx`

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    return current_conv, match, mistake_type
+
+
+def evaluate_gpt(ds: List[dict], openai_parser: OpenAIResponseParser):
```
use a better name than ds (use compelete names)

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    num_correct = 0
+    total_count = 0
+    results = []
+    pbar = tqdm(total=len(ds), desc="Evaluating GPT4..")
```
```suggestion
    pbar = tqdm(total=len(ds), desc="Evaluating GPT4 ...")
```

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+    anyscale_accuracy = 0
+    total_count = 0
+    results = []
+    pbar = tqdm(total=len(ds), desc="Evaluating Finetuned Model..")
```
```suggestion
    pbar = tqdm(total=len(ds), desc="Evaluating Finetuned Model ...")
```

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+
+        if openai_is_match:
+            num_correct += 1
+    print("GPT4 accuracy: ", num_correct / total_count)
```
in jupyter notebooks, don't print stuff in functions and only print them in the notebooks

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,163 @@
+        anyscale_messages = example["anyscale_messages"]
+        anyscale_conv, anyscale_is_match, mistake_type = parse_and_eval(anyscale_parser, anyscale_messages, example["expected_responses"])
+        pbar.update(1)
+        if anyscale_conv is None: # skip if api errors out
+            anyscale_is_match = None
+        else:
+            total_count += 1
```
unify the return types. Don't mix up None with empty string, etc. Or at least define the interface more consistently.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/function_extraction_utils.py:
```diff
@@ -0,0 +1,65 @@
+"""
+Function call json extraction utils.
+Reference: https://gist.github.com/kouroshHakha/6cfbe2bf4aaafc5db733d408044f9902#file-create_test_dataset-py
```
don't mention the random reference in the template :)

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/function_extraction_utils.py:
```diff
@@ -0,0 +1,65 @@
+    brace_count = 0
+    start = None
+
+    for i, char in enumerate(jsonl_string):
+        if char == '{':
+            if brace_count == 0:
+                start = i
+            brace_count += 1
+        elif char == '}':
+            brace_count -= 1
+            if brace_count == 0:
+                end = i + 1
+                json_strs.append(jsonl_string[start:end])
```
use regex or sth less adhoc and more systematic?

kouroshHakha replied:
Ask AI to improve your code.

SumanthRH replied:
Is this not systematic? i keep track of open braces and then when all are closed, I extract json from the first open brace. It will work with all sorts of special characters etc. 


Regex was my go-to but I've encountered issues with nested JSONs. The format here is that  you have a string which contains one or more jsons separated by commas. You can have nested jsons separated by commas and regex won't work then (atleast the patterns I tried).  The code here is a bit slower but will be correct.

kouroshHakha replied:
This code is not maintainable. You are adding complexity in a place whose concern is not parsing json. So the preference is to find a one / two liner usining an existing popular library to do the parsing. Say tomorrow this part of the code turns out to be problematic, it will be a lot harder for someone to resolve the issue then. So my question to you is what is the complex jsonl_string that you couldn't use regex for?

SumanthRH replied:
A part of the problem was that I was using the same function in some places to process Glaive's system message (where the jsons are space separated) and our formatted tool calls (where they are comma separated). I've refactored the whole thing now to have use to separate the two kinds of formats. I've also used one common function to process our tool info formatting (which is "[prefix_content] <tag1> [tool info] <tag2>") for tool lists, tool calls, tool results, etc. and then leave further parsing to `json`.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/function_extraction_utils.py:
```diff
@@ -0,0 +1,65 @@
+def get_tool_calls_from_response(assistant_content):
+    assistant_content = assistant_content.strip()  # remove trailing whitespaces
+    escaped_tool_call_tags = [re.escape(tag) for tag in TOOL_CALL_TAGS]
+    if assistant_content.startswith(TOOL_CALL_TAGS[0]):
```
what is TOOL_CALL_TAGS[0]??

kouroshHakha replied:
Never hardcode things like this.

kouroshHakha replied:
Let's do some excercise (it will be a good one)

For each function you have defined, make high level docsring with all types hints and everything. Then feed the docstring and signature into AI and see how well it implemented it (catch its mistakes ofc, but I generally found its style much better than mine it will also give you better ideas on how to make code cleaner and easier to follow, because it was trained to do that). Overtime you'll pick up the right style.

SumanthRH replied:
> Never hardcode things like this. 

"TOOL_CALL_TAGS" are the special indicators used during preprocessing ("[TOOL_CALLS]" and "[/TOOL_CALLS]") . Our post-processing code for the responses has to be pre-processing specific! There's no way to extract assistant response jsons if I don't know what tool call tags I used (assistant response can sometimes have valid json, but not a tool call).  

This is not buried deep within the code btw. It's at the top of [preprocessing.py](https://github.com/anyscale/templates/blob/b84ccda3ba3d22faa8a478c75d40526625ed3158/templates/fine-tune-function-calling/fc_utils/preprocessing.py#L11). This is one code file that has to have a dataset specific implementation. 

WDYT?

SumanthRH replied:
> For each function you have defined, make high level docsring with all types hints and everything. Then feed the docstring and signature into AI and see how well it implemented it (catch its mistakes ofc, but I generally found its style much better than mine it will also give you better ideas on how to make code cleaner and easier to follow, because it was trained to do that). Overtime you'll pick up the right style.

Interesting. Let's see :)

kouroshHakha replied:
I see let's do something like this:

```
class ToolCallsTag(Enum):
       start = "[TOOL_CALLS]"
       end = "[/TOOL_CALLS]"
```

SumanthRH replied:
I've put a higher-level NamedTuple called IndicatorTags and used that to instantiate our specific tool list, tool call and tool result tags.

kouroshHakha reviewed (commented):
Submitting half of my comments (more to come):

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
````diff
@@ -0,0 +1,419 @@
+
+```python
+# complete preprocessing step
+processed_ds = preprocess(ray_ds) 
````
Use better names here:

so you'll be doing `anyscale_ds = openai_to_anyscale(ray_ds)`
Let's call the first one `openai_ds = glaive_to_openai(ray_ds)`

So these will be the two public functions. Everything else inside of it should be private function indicated by `_` in front of the name. e.g. `_filter_bad_examples`.

Also why is filtering not applied at when creating the openai dataset? It should be like the following flow:

```
raw_glaive_ds -> openai fmt -> filtered_openai -> anyscale fmt
```

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/preprocessing.py:
```diff
@@ -0,0 +1,254 @@
+    return example
+
+
+def filter_func(example: Dict[str, Any]) -> bool:
```
returns bool but you are returning int.

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
```diff
@@ -0,0 +1,419 @@
+Follow the instructions to run your fine-tuning job.
+
+## Step 2(b): Fine-tuning through serverless endpoints
+First, obtain your credentials from the [Anyscale platform](https://console.anyscale.com/v2/) and upload the training and validation files in the [fine-tuning tab](https://console.anyscale.com/v2/fine-tuning?fine-tuning-tab=files). Make a note of the file IDs for each. This will be passed to the fine-tuning job.
```
They don't need to upload the files separately from the finetuning job. They can upload data at the time of finetuning submision.

kouroshHakha replied:
so instead of going first under files tab, they can just go to jobs tab and create and upload at the same time.

kouroshHakha replied:
So here you are kinda mixing UI and SDK. 

We can just make it SDK only. You can include code for data upload here as well. And just punt on file upload via UI.

SumanthRH replied:
Ah... I actually missed this part of the SDK! Have updated the upload instructions now.

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
````diff
@@ -0,0 +1,419 @@
+
+
+```python
+# to be run only if you finetuned on the Anyscale platform
+ANYSCALE_API_KEY="service-api-key-here"
+ANYSCALE_API_BASE="service-url-here" 
+if not ANYSCALE_API_BASE.endswith("/"):
+    ANYSCALE_API_BASE += "/"
+ANYSCALE_API_BASE += "v1"
````
use fstring

kouroshHakha replied:
just comment on what they should paste (don't use the endswith `/` stuff.) Maybe a screen shot with highlights around the text that is important.

SumanthRH replied:
Yup added.

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
```diff
@@ -0,0 +1,419 @@
+
+### (Optional) Try out the model via Playground
+
+You can try out your new model in the Playground: https://console.anyscale.com/v2/playground . In the model dropdown, you should be able to see your finetuned model as shown below
```
ok playground experience is limited to sdk finetuned models only, so make it clear.

SumanthRH replied:
Good idea. I just kept it to be a subsection for the serverless endpoints finetuning, but added an explicit note.

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
```diff
@@ -0,0 +1,419 @@
+
+Evaluation of function calling capability is non-trivial, given that we're looking to extract structured data from an inherently unpredictable and unstructured stream of text. We will use the following simple evaluation strategy: The models are evaluated on the accuracy metric and their responses are graded as accurate if their response for each assistant entry in the conversation is correct. An  ...
+1. In case the ground truth response contains no function call, then the model's response should also not have a function call. 
+2. In case the ground truth response contains a function call, then the model's response should also have a function call. We do not check for the content of the response here. The assistant function call should further have the correct function name and the correct function arguments. 
```
> We do not check for the content of the response here.
This is confusing? I know you mean the "content" in the message but instead of this just say we only check the name and arguments of the function call to ensure correctness.

SumanthRH replied:
Have fixed the wording now.

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
````diff
@@ -0,0 +1,419 @@
+
+```python
+# initialize parsers
+openai_parser = OpenAIResponseParser(api_key=OPENAI_API_KEY, api_base="https://api.openai.com/v1", model="gpt-4", tool_call_tags=TOOL_CALL_TAGS)
````
"https://api.openai.com/v1" use a const definition above where you define the ANYSCALE_API_BASE?

SumanthRH replied:
Have added this along with the API keys section above.

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
````diff
@@ -0,0 +1,419 @@
+
+```python
+# evaluate both models and plot the results
+results_gpt, accuracy_gpt = evaluate_model(modified_ds, openai_parser, "gpt")
````
separate teh evaluate_model in different blocks, because this takes time.

kouroshHakha reviewed (commented):
Notebook feedback:
I have left reviews on the README.md. Please make the changes to the notebook such that the generated README will get fixed. 

Code architecture feedback:
- Let's revise the data flow: 
The initial mapper currently goes from glaive dataset to an arbitray format that is neither openai neither anyscale. What I would do? I would create an API that looks like:  ray_ds.map(glaive_to_openai) and implement glaive_to_openai to be the dataset format compatible with openAI. Right now your assistant content includes [TOOL_CALL] tags which is anyscale specific while you also have a tool role in your messages. The issue is that you are combining structured with unstructured together here, which means you have to carry the complexity around all over the place. You can avoid it by being smarter at the beginning.

Let's also make it clear (with a graph) that we are doing: 

glaive -> openAI -> Anyscale

in terms of formatting. 

- I see that you wrote the pprint_example in a way that accept keys (probably because the example that you pass into the pprint includes other things you don't want to print). The reason for this is that your initial_mapper is not implemented properly. Why are you keeping the old keys in the return dict from initial_mapper?

Style feedback:

- Make sure your IDE wraps lines at 80 characters. Then go through each file and see if wrapping has happened. If so please break the lines

You'll find other specific comments inside the code.

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,207 @@
+
+
+def evaluate_model(dataset: List[Dict[str, Any]], parser: ResponseParser, model: str) -> Tuple[List[Dict[str, Any]], float]:
+    """
+    Evaluates the given model on the test dataset. The function returns a list of results and the accuracy.
+
+    Each result entry is a dictionary containing the following:
+        - correct: Boolean indicating if the model's response matches the ground truth
+        - invalid_entry: Boolean indicating if the entry is invalid. This can happen if the api call failed for some reason (max retries reached, etc.). Such entries are skipped in accuracy calculation.
+        - mistake_type: String indicating the type of mistake made by the model if any
+        - example: The original example from the dataset
+        - conv: The full conversation with the model's actual response.
+
+    Args:
+        dataset: List of examples to evaluate
+        openai_parser: ResponseParser object
+
+    Returns:
+        results: List of results
+    """
```
Note that in docstring above, first line is a short (less than 60-70 character) description.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/preprocessing.py:
```diff
@@ -0,0 +1,254 @@
+        keys = example.keys()
+    pprint_str = ""
+    # ANSI escape code for blue, green, red, yellow and magenta colors
+    blue = "\033[94m"
+    reset = "\033[0m"
+    green = "\033[92m"
+    red = "\033[91m"
+    magenta = "\033[95m"
+    yellow = "\033[93m"
```
done! have also taken care of different data formats now so that this looks much nicer.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/preprocessing.py:
```diff
@@ -0,0 +1,254 @@
+    return ray_ds
+
+
+def pprint_example(example: Dict[str, Any],keys: List[str] = []) -> None:
```
Hmm yes this was not meant to be purely openai. The flow I had in my mind was to follow the dataset's format and then change it to something like the openai format and then bring to anyscale. It's now changed to exactly the openai format and then our format.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/preprocessing.py:
```diff
@@ -0,0 +1,254 @@
+    return messages
+
+
+def final_mapper(example: Dict[str, Any]) -> Dict[str, Any]:
```
Yes this is actually better. I took the path of least resistance first and the intermediate dataset doesn't add much information. Have changed `initial_mapper` to `glaive_to_openai` now.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/function_extraction_utils.py:
```diff
@@ -0,0 +1,95 @@
+def extract_jsons(string: str) -> List[Dict[str, Any]]:
+    """
+    Extracts JSON objects from a string containing one or more JSONs.
+    Example: "[{"name": "weather", "arguments": {\"location\": \"New York\"}}]"
```
Hmm the problem I had was that I wanted to use this function in both places: for the glaive system message and for extracting tool calls from assistant responses. In the case of the former, the jsons are separated by a space and in the latter, I wanted to have them comma separated (more standard). This made things complicated with regex. I've now stuck to using regex for the system message through `extract_functions_from_system_msg` and then made two sets of parsers for assistant calls in glaive and openai/anyscale responses.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/preprocessing.py:
```diff
@@ -0,0 +1,254 @@
+            if GLAIVEAI_TEMPLATE["tool_call_prefix"] in assistant_content:
+                # make function call a list and add tags
+                assistant_content = assistant_content.replace(
+                    GLAIVEAI_TEMPLATE["tool_call_prefix"], f"{TOOL_CALL_TAGS[0]} ["
```
Yes I actually wrote that this is an Openai-like conversation format.  The flow I had in my mind was to stay close to the dataset's format and then change it to something _like_ the openai format and then bring to anyscale. It's now changed to exactly the openai format and then small modifications for our format.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/plot_utils.py:
```diff
@@ -0,0 +1,61 @@
+ # 10 colors for different flags
+COLORS = ['salmon', 'deepskyblue', 'springgreen', 'orange', "gold", "limegreen", "turquoise", "blueviolet", "darkred", "darkorange"]
+
+def plot_results(results_finetuned, results_gpt):
```
On it! Will standardize a result format as well. This plotting function was initially a second-class member of the utilities package, so I'll try to make this neater.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,207 @@
+    """
+    if model not in ["gpt", "finetuned"]:
+        raise ValueError(f"Model must be either 'gpt' or 'finetuned', got {model}")
+    num_correct = 0
+    total_count = 0
```
Hmm I wanted to avoid wasting memory, but agreed that the above approach is more pythonic and looks cleaner. Have switched to this.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,207 @@
+        if conv is None:
+            invalid_entry = True
+        else:
+            # valid, increment total count
```
Taking care of this now!

SumanthRH commented on templates/fine-tune-function-calling/README.md:
```diff
@@ -0,0 +1,419 @@
+
+
+# Table of Contents
+1. [Data Preprocessing](#step-1-data-preprocessing)
+2. [Finetuning](#step-2-fine-tuning)
+3. [Serving](#step-3-serving)
+4. [Evaluation](#step-4-evaluation)
```
In terms of the flow, this doesn't make much sense for me. Since we want the user to evaluate the model they fine-tuned in Step 2, serving should come before evaluation (otherwise what endpoint will `ANYSCALE_API_BASE` be)

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,207 @@
+    def values():
+        return [item.value for item in Mistakes]
+
+def check_match(response: Dict[str, Union[str, ToolCallType]], ground_truth: Dict[str, Union[str, ToolCallType]]) -> Tuple[bool, Mistakes]:
```
Done!

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/eval_utils.py:
```diff
@@ -0,0 +1,207 @@
+    Checks if the tool call is in the correct format.
+    """
+    for tool_call in tool_calls:
+        if "name" not in tool_call or "arguments" not in tool_call or not isinstance(tool_call["arguments"], dict) or not isinstance(tool_call["name"], str):
```
Done! Since this function also enforces a specific tool format, I've moved it to my data formatting/constants util file `data_format.py`

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/function_extraction_utils.py:
```diff
@@ -0,0 +1,95 @@
+
+
+# define our custom type for a tool call
+ToolCallType: TypeAlias = Dict[str, Union[str, Dict[str, str]]]
```
Good catch! Removed the type hint

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/preprocessing.py:
```diff
@@ -0,0 +1,325 @@
+            if len(tools):
+                tools_str = json.dumps(tools)
+                tools_str_with_tags = (
+                    f"{TOOL_LIST_TAGS[0]} {tools_str} {TOOL_LIST_TAGS[1]}"
```
TOOL_LIST_TAGS.start and end?

SumanthRH replied:
Yes done. I was still cleaning up!

kouroshHakha commented on templates/fine-tune-function-calling/fc_utils/preprocessing.py:
```diff
@@ -0,0 +1,325 @@
+    FunctionCallNotFoundError,
+    DatasetFormat,
+)
+from fc_utils.data_format import *
```
never do `from x import *` always be explicit about imports.

SumanthRH replied:
Hmm that rule can have exceptions.  Especially with a case like here where one file has mostly constants and classes all used in the other file (preprocessing.py). I've switched to the explicit import anyways (though it's ugly and unnecessary!).

kouroshHakha commented on templates/fine-tune-function-calling/README.md:
````diff
@@ -0,0 +1,478 @@
+If you notice, each sample has two entries: system and chat. This dataset is already formatted in specific way (e.g. using USER, \<|endoftext|\> and other tokens). To enable fine-tuning on various open source models we need to convert each row to a more general format like the OpenAI chat format, which is the preferred format for fine-tuning instruction-tuned models on Anyscale ([dataset format g ...
+
+
+```python
+import fc_utils.preprocessing
+import importlib
+importlib.reload(fc_utils.preprocessing)
+from fc_utils.preprocessing import pprint_example
+```
````
what are these? can be removed?

SumanthRH replied:
I was still cleaning this up! Removed now with all new changes incorporated.

kouroshHakha reviewed (commented):
Really good. A few comments left.

SumanthRH commented on templates/fine-tune-function-calling/fc_utils/test_utils.py:
```diff
@@ -0,0 +1,182 @@
+    tool_result_tags: IndicatorTags,
+    tool_list_tags: IndicatorTags,
+    format: DatasetFormat,
+):
```
Outdated! Was in the middle of pushing these fixes

kouroshHakha commented:
I'd like add one thing on top of Goku's comments:

> I'm not sure how unavoidable this since the dataset is quite involved is but it was a bit difficult to follow the evaluation workload: evaluate_model --> parse_and_eval --> get_parsed_response --> get_completion --> check_match. Is there a way to consolidate this?

Can we create a flow-chart in the readme that shows the process so that the reader can understand the architecture of the design if they want to dive in the code.

akshay-anyscale commented on templates/fine-tune-function-calling/README.ipynb:
```diff
@@ -0,0 +1,761 @@
+    "\n",
+    "## Step 3(a): Finetuned on the Anyscale Platform\n",
+    "\n",
+    "Make a note of the final checkpoint after fine-tuning (this should be the last line in the logs). You can now spin up the \"Deploy LLMs\" template which has all the instructions and required dependencies to serve your finetuned model efficiently. You will find the tutorials on [serving LoRA models](https://github.com/anyscale/templates/blob/main/templates/endpoints_v2/examples/lora/DeployLor ...
```
Can you update this based on the new flow in the template with a generate config script?

SumanthRH replied:
Hmm this won't be a standalone template so just I'm directing them to the serve template - that will have all the instructions. So wouldn't it be better to keep this as is?

kouroshHakha replied:
+1 to @SumanthRH

SumanthRH commented:
@GokuMohandas @kouroshHakha Addressing your comments here:

> I'm not sure how unavoidable this since the dataset is quite involved is but it was a bit difficult to follow the evaluation workload: evaluate_model --> parse_and_eval --> get_parsed_response --> get_completion --> check_match. Is there a way to consolidate this?

> Can we create a flow-chart in the readme that shows the process so that the reader can understand the architecture of the design if they want to dive in the code.

I hope the eval flowchart clarifies why we have the current setup!

> Also in the error analysis, would it worth showing show a base model would perform? I can appreciate that the fine-tuned model and GPT-4 both perform roughly the same number of mistakes, but I'd love to see how bad the original base model was.

Good idea, I've added this! The base model system prompt is not completely optimized, but initial results show a big gap:

![image](https://github.com/anyscale/templates/assets/39546518/758dbc57-fcca-4a53-939a-4d152e934a5d)

I've also added some small explainers for how I'm evaluating the base model, let me know if it makes sense (Also, you can take a look at the system prompt here: https://github.com/anyscale/templates/blob/df456a0ccc4ce32bc016c841dae90fda5e52992c/templates/fine-tune-function-calling/fc_utils/data_format.py#L18 

> Also when we serve these models, do you have to apply the chat template manually or at least ensure that theprompt_format is correct in our model_config_<timestamp> file? I may have missed this if you do it somewhere already.


We're not touching this at all! Data processing is operating at the OpenAI-chat interface so I'm only figuring out how the "user" "system" responses (and "assistant" responses in case of training) look like. Applying the chat template is left to our internal fine-tuning and serving code (which is consistent).

kouroshHakha commented:
I have not spelled check things again. The plots / figures add much better narration impact on the reader. So approved.
